### PR TITLE
Issue 2252: Refactor the referrer hiding.

### PR DESCRIPTION
**Pull request by iefremov**
Fix https://github.com/brave/brave-browser/issues/2252

Current solution for hiding the referer string requires readirecting literally
all requests for which the referrer was modified. This approach conflicts
with the fact that certain requests are not allowed to be redirected (i.e.
CORS preflight requests). That said, current approach breaks a lot of sites.

To refactor this we move the referrer hiding code to the earlier stage,
where navigation requests are created and started. This is a convenient
place to patch both renderer and browser initiated navigations and utilize
ContentBrowserClient interface on the browser thread.

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Windows
  - [ ] macOS
  - [x] Linux
- Verified that these changes pass automated tests (`npm test brave_unit_tests && npm test brave_browser_tests`) on
  - [ ] Windows
  - [ ] macOS
  - [x] Linux
- [x] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

### CODE REVIEW

**Comment by iefremov** [comment]
TODO:
 * Add test against the origin CORS problem (chromium has one, but it is buried deeply in blink tests infrastructure)
 * Add unittests against referrer blocking function instead of removed ones
 * Possibly add referrer hiding call to `NavigationRequest::OnRequestRedirected`

File: browser/brave_content_browser_client.cc [hunk 1]
```diff
@@ -259,6 +260,39 @@ void BraveContentBrowserClient::AdjustUtilityServiceProcessCommandLine(
...
+          CONTENT_SETTINGS_TYPE_PLUGINS,
+          brave_shields::kBraveShields);
+   brave_shields::ShouldSetReferrer(allow_referrers, shields_up,
+                                    referrer->url, document_url, request_url,
+                                    request_url.GetOrigin(),
+                                    referrer->policy,
+                                    referrer);
+}
+
+
```
Comment (yrliou) [inline browser/brave_content_browser_client.cc]: nit: Please remove the extra line.

File: components/brave_shields/browser/brave_shields_util.cc [hunk 2]
```diff
@@ -48,6 +51,32 @@ bool GetDefaultFromResourceIdentifier(const std::string& resource_identifier,
...
+  if (setting == CONTENT_SETTING_DEFAULT) {
+    return GetDefaultFromResourceIdentifier(resource_identifier, primary_url,
+                                            secondary_url);
+  }
+  return setting == CONTENT_SETTING_ALLOW;
+}
+
+}  // namespace
+
+
```
Comment (yrliou) [inline components/brave_shields/browser/brave_shields_util.cc]: nit: remove the extra line

File: browser/net/brave_site_hacks_network_delegate_helper_unittest.cc [hunk 3]
```diff
@@ -215,88 +215,4 @@ TEST_F(BraveSiteHacksNetworkDelegateHelperTest, NOTUAWhitelistedTest) {
   });
 }
 
-TEST_F(BraveSiteHacksNetworkDelegateHelperTest, ReferrerPreserved) {
```
[the remark this answers is not in the document]
Comment (iefremov) [reply browser/net/brave_site_hacks_network_delegate_helper_unittest.cc]: done

File: browser/brave_content_browser_client.cc (hunk 1 above)
Comment (iefremov) [reply browser/brave_content_browser_client.cc]: done

File: components/brave_shields/browser/brave_shields_util.cc (hunk 2 above)
Comment (iefremov) [reply components/brave_shields/browser/brave_shields_util.cc]: done

**Pull request merged**